modules/globebrowsing/cache/memoryawaretilecache.cpp

```diff
@@ -32,51 +35,204 @@ namespace globebrowsing {
+        it++)
+    {
+        it->second.first->reset();
+        it->second.second->clear();
+    }
+}
+
+void MemoryAwareTileCache::createDefaultTextureContainers() {
+    for (int id = 0; id < layergroupid::NUM_LAYER_GROUPS; id++) {
+        TileTextureInitData initData = LayerManager::getTileTextureInitData(layergroupid::ID(id));
+        assureTextureContainerExists(initData);
+    }
+}
+
+void MemoryAwareTileCache::assureTextureContainerExists(
+    const TileTextureInitData& initData)
+{
+    TileTextureInitData::HashKey initDataKey = initData.hashKey();
+    if (_textureContainerMap.find(initDataKey) == _textureContainerMap.end()) {
+        // For now create 500 textures of this type
+        _textureContainerMap.emplace(initDataKey,
+            TextureContainerTileCache(
+                std::make_unique<TextureContainer>(initData, 500),
+                std::make_unique<TileCache>(std::numeric_limits<std::size_t>::max())
+            )
+        );
+    }
+}
+
+void MemoryAwareTileCache::setSizeEstimated(size_t estimatedSize) {
+    ghoul_assert(_textureContainerMap.size() > 0, "Texture containers must exist.");
+    size_t sumTextureTypeSize = 0;
+    for (TextureContainerMap::const_iterator it = _textureContainerMap.cbegin();
+        it != _textureContainerMap.cend();
+        it++)
+    {
+        sumTextureTypeSize +=
+            it->second.first->tileTextureInitData().totalNumBytes();
+    }
+    size_t numTexturesPerType = estimatedSize / sumTextureTypeSize;
+    resetTextureContainerSize(numTexturesPerType);
+}
+
+void MemoryAwareTileCache::resetTextureContainerSize(size_t numTexturesPerTextureType) {
+    _numTextureBytesAllocatedOnCPU = 0;
+    for (TextureContainerMap::const_iterator it = _textureContainerMap.cbegin();
+        it != _textureContainerMap.cend();
+        it++)
+    {
+        it->second.first->reset(numTexturesPerTextureType);
+        it->second.second->clear();
+    }
 }
 
 bool MemoryAwareTileCache::exist(ProviderTileKey key) const {
-    std::lock_guard<std::mutex> guard(_mutexLock);
-    return _tileCache.exist(key);
+    for (TextureContainerMap::const_iterator it = _textureContainerMap.cbegin();
+        it != _textureContainerMap.cend();
+        it++)
+    {
+        if(it->second.second->exist(key)) {
+            return true;
+        }
+    }
+    return false;
 }
 
 Tile MemoryAwareTileCache::get(ProviderTileKey key) {
-    std::lock_guard<std::mutex> guard(_mutexLock);
-    return _tileCache.get(key);
+    for (TextureContainerMap::iterator it = _textureContainerMap.begin();
+        it != _textureContainerMap.end();
+        it++)
+    {
+        if(it->second.second->exist(key)) {
+            return it->second.second->get(key);
+        }
+    }
+    ghoul_assert(false, "");
 }
 
-void MemoryAwareTileCache::put(ProviderTileKey key, Tile tile) {
-    std::lock_guard<std::mutex> guard(_mutexLock);
-    _tileCache.put(key, tile);
+ghoul::opengl::Texture* MemoryAwareTileCache::getTexture(
```

Comment:
In my opinion it should be a pointer to be clearer about that it is an object that does not get copied when getting it.